From c8446556320874d5c4da2a22001f672b6473140f Mon Sep 17 00:00:00 2001
From: Ricardo Neri <ricardo.neri@ti.com>
Date: Tue, 28 Jun 2011 23:12:07 -0500
Subject: [PATCH 1/2] OMAP4: HDMI: Restore previous state of ACR WA after SYSM3 crash

After recovering from a SYSM3 crash, restore the ACR WA to the state
it has before the crash.

Change-Id: I77b4e22bbe8b5fae76877c328b527e4e0ae3f8ec

Signed-off-by: Ricardo Neri <ricardo.neri@ti.com>
---
 arch/arm/plat-omap/hdmi_lib.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

Index: kernel/arch/arm/plat-omap/hdmi_lib.c
===================================================================
--- kernel.orig/arch/arm/plat-omap/hdmi_lib.c	2011-07-01 18:11:25.979648195 +0530
+++ kernel/arch/arm/plat-omap/hdmi_lib.c	2011-07-01 18:11:37.099653746 +0530
@@ -1522,6 +1522,7 @@
 {
 	int status = 0;
 	u16 *proc_id = (u16 *)v;
+	static u32 wa_state_last = HDMI_WA_STATE_NOTIFY_EVENT_NOTREG;
 
 	switch ((int)val) {
 	case IPC_START:
@@ -1533,6 +1534,10 @@
 				mutex_lock(&hdmi.wa_state_lock);
 				hdmi.wa_state = HDMI_WA_STATE_STOPPED;
 				mutex_unlock(&hdmi.wa_state_lock);
+				if (wa_state_last == HDMI_WA_STATE_STARTED ||
+					wa_state_last == HDMI_WA_STATE_TRANSIT_TO_START)
+					hdmi_lib_start_acr_wa();
+
 			}
 		}
 		return status;
@@ -1541,6 +1546,10 @@
 			status = notify_unregister_event(SYS_M3, 0,
 				HDMI_AUDIO_WA_EVENT_ACK, (notify_fn_notify_cbck)
 				hdmi_notify_event_ack_func,	(void *)NULL);
+				mutex_lock(&hdmi.wa_state_lock);
+				wa_state_last = hdmi.wa_state;
+					mutex_unlock(&hdmi.wa_state_lock);
+
 			if (status == NOTIFY_S_SUCCESS) {
 				mutex_lock(&hdmi.wa_state_lock);
 				hdmi.wa_state =
